### PR TITLE
Test cases

### PR DESCRIPTION
**Pull request by emanuel-skrenkovic**
https://github.com/fsnotify/fsevents/pull/62#issuecomment-2091766682

PR to tests fsnotify style test scripts branched the main branch, as discussed in link above.

### CODE REVIEW

**Comment by arp242** [comment]
It seems the main bit that's missing here is that it doesn't actually print and parse the list of events? It seems always 0:

            have:
            	0        /
            	0        /file
            want:
            	0        /file

Here's what I would do:

1. Fix that.
2. Just remove all tests, as the behaviours seem too different.
3. Start adding them back one-by-one, fixing up the Output, removing the cross-platform bits that are no longer needed.

You can also use a different approach if that works better for you, but the end result should be something that has tests for at least a bunch of common situations: "make a new directory", "remove watched directory", etc. etc.

**Comment by emanuel-skrenkovic** [comment]
Added some tests now. The tests that are here run and pass.

Behaviour is different from fsnotify, and I've mostly just replaced the test asserts with what is happening. These aren't really correctness tests at this point, just regression.

I also haven't really had anything to compare against, as there are not too many other tests here. That is something to keep in mind.

I was watching what was happening for each test though, and the events emitted do make sense with regards to what the test does, and what I could find on how fsevents work on mac os.

File: helpers_test.go [hunk 1]
```diff
@@ -0,0 +1,729 @@
...
+				t.Fatalf("line %d: %q requires exactly %d argument (have %d: %q)",
+					c.line, c.cmd, n, len(c.args), c.args)
+			}
+		}
+	)
+loop:
+	for _, c := range cmds {
+		c := c
+		switch c.cmd {
+		case "skip", "require":
```
Comment (arp242) [inline helpers_test.go:595]: I think this command can be removed?

File: helpers_test.go [hunk 2]
```diff
@@ -0,0 +1,729 @@
...
+	"ItemRemoved":       ItemRemoved,
+	"ItemInodeMetaMod":  ItemInodeMetaMod,
+	"ItemRenamed":       ItemRenamed,
+	"ItemModified":      ItemModified,
+	"ItemFinderInfoMod": ItemFinderInfoMod,
+	"ItemChangeOwner":   ItemChangeOwner,
+	"ItemXattrMod":      ItemXattrMod,
+	"ItemIsFile":        ItemIsFile,
+	"ItemIsDir":         ItemIsDir,
+	"ItemIsSymlink":     ItemIsSymlink,
```
Comment (arp242) [inline helpers_test.go:38]: Can these just be "created", "removed", "isdir", etc. without the "item" every time? It's very repetitive and "noisy" in the test files.

File: helpers_test.go [hunk 3]
```diff
@@ -0,0 +1,729 @@
...
+				continue
+			}
+			t.Fatalf("newEvents: line %d: needs 2 or 4 fields: %s", no+1, line)
+		}
+
+		var resultFlags EventFlags
+
+		flags := strings.Split(fields[0], "|")
+		for _, f := range flags {
+			resultFlags = resultFlags.set(eventFlagsPossible[f])
```
Comment (arp242) [inline helpers_test.go:454]: I don't think this is case-insensitive? It should be: "created", "CREATED", and "Created" should all work. No reason to be pedantic about capitals here.

**Comment by arp242** [comment]
> Behaviour is different from fsnotify, and I've mostly just replaced the test asserts with  what is happening. These aren't really correctness tests at this point, just regression. I also haven't really had anything to compare against, as there are not too many other tests here. That is something to keep in mind.

Yeah, that's fine. The important bit here is to ensure the behaviour doesn't radically change and that it doesn't panic.

Also, can you add a test like so (completely untested code for illustrative purposes):

	func TestMany(t *testing.T) {
		tmp := t.TempDir()
		var wg sync.WaitGroup
		wg.Add(300)
		for i := 0; i<100; i++ {
			i := i
			go func() { defer wg.Done(); echoAppend(fmt.Sprintf(%s/file-%d", tmp, i)) }()
			go func() { defer wg.Done(); mkdir(fmt.Sprintf(%s/dir-%d", tmp, i)) }()
			go func() { defer wg.Done(); echoAppend(fmt.Sprintf(%s/dir-%d/file-%d", tmp, i)) }()
		}
		wg.Wait()

		wg.Add(200)
		for i := 0; i<100; i++ {
			i := i
			go func() { defer wg.Done(); remove(fmt.Sprintf(%s/file-%d", tmp, i)) }()
			go func() { defer wg.Done(); removeAll(fmt.Sprintf(%s/dir-%d", tmp, i)) }()
		}
		wg.Wait()

		// Check if you have 100 writes for new files, 100 removes, etc.
	}


That is, create 100 files and directories simultaneously in a watched directory, then remove them, and ensure we've got the right events.

After that and the minor comments it should be good to go.

File: helpers_test.go (hunk 2 above)
Comment (emanuel-skrenkovic) [reply helpers_test.go:38]: Hmm, I can change it, it is no problem, but I made it the same as the name of the const to immediately know what is being referred to here. Also makes it grep-able without the extra step. 
It is a bit of a crappy solution admittedly.

Don't have a strong opinion though. If you'd like, I can change it into more human readable names.

Comment (arp242) [reply helpers_test.go:38]: It's not the biggest thing in the world; you can leave it if that's what you prefer.

**Comment by arp242** [comment]
Seems good; thanks!

**Pull request merged**